compat/src/test/scala/test/scala/collection/LazyListLazinessTest.scala

```diff
@@ -776,24 +790,45 @@ class LazyListLazinessTest {
     assertRepeatedlyLazy(factory)
   }
 
-  @Test
-  def `#:: properlyLazy`(): Unit = {
-    val factory = lazyListFactory { init =>
+  private def genericCons_properlyLazy(cons: (=> Int, => LazyList[Int]) => LazyList[Int]): Unit = {
+    val headInitFactory = lazyListFactory { init =>
       def gen(index: Int): LazyList[Int] = {
         def elem(): Int = { init.evaluate(index); index }
         if (index >= LazinessChecker.count) LazyList.empty
-        // else elem() #:: gen(index + 1)
-        else gen(index + 1).#::(elem())
+        else cons(elem(), gen(index + 1))
       }
 
       gen(0)
     }
-    assertRepeatedlyLazy(factory)
+    assertRepeatedlyLazy(headInitFactory)
+
+    val tailInitFactory = lazyListFactory { init =>
+      def gen(index: Int): LazyList[Int] = {
+        if (index >= LazinessChecker.count) LazyList.empty
+        else {
+          init.evaluate(index)
+          cons(index, gen(index + 1))
+        }
+      }
+
+      LazyList.empty lazyAppendedAll gen(0) // prevent initial state evaluation
+    }
+    assertRepeatedlyLazy(tailInitFactory)
+  }
+
+  @Test
+  def cons_properlyLazy(): Unit = {
+    genericCons_properlyLazy(LazyList.cons(_, _))
+  }
+
+  @Test
+  def `#::_properlyLazy`(): Unit = {
+    genericCons_properlyLazy(_ #:: _)
```

Comment:
you'll need to change this because the arguments to right-associative operators are not lazy in 2.12-

```suggestion
    // genericCons_properlyLazy(_ #:: _)
    genericCons_properlyLazy((hd, tl) => tl.#::(hd))
```